research/train.py

```diff
@@ -113,29 +126,49 @@ def train(cfg: DictConfig) -> None:
     train_loader, val_loader = get_dataloaders(cfg, num_workers=cfg.data_loader_workers)
 
     # Training loop logic
-    trainer = create_supervised_trainer(model, cfg, device=device)
-
+    trainer = create_training_loop(model, cfg, device=device)
+    trainer.logger = setup_logger(name="trainer")
+    
     # Evaluation loop logic
-    evaluator = create_supervised_evaluator(model, cfg, device=device)
-
-    trainer.logger = setup_logger("trainer")
-    evaluator.logger = setup_logger("evaluator")
-
+    evaluator = create_evaluation_loop(model, cfg, device=device)
+    evaluator.logger = setup_logger(name="evaluator")
+    
     ########################################################################
     # Callbacks
     ########################################################################
 
-    # When epoch completes, run evaluator engine on val_loader, then log ["accuracy", "nll"] metrics to file and stdout.
+    #!!!! Required. Do not change. !!!!#
+    trainer.add_event_handler(Events.STARTED, setup_file_pointers)
+    evaluator.add_event_handler(Events.STARTED, setup_file_pointers)
+
+    # Perform various log operations on the "trainer" engine output every 50 iterations
     trainer.add_event_handler(
```

Comment:
These are currently added manually, but eventually we might want to move the eventhandler adding in some config setup:
```python
def attach_events(engine, config):
    for event in config[event]:
        engine.add(event[name], event[onEventTriggerHandler])
```

Reply:
So you're suggesting logging as config? I'm concerned that's too much abstraction. Because we're going to extend beyond just log operations, we also want action operations at some point.

Reply:
So some actions you might have are like use a different training pattern, learning rate, optimizer or any other action you would want at a certain event, wouldn't that be a hyperparameter? The way I view configs is that they are training settings. Technically logging isn't a training setting, but if we allow actions (which we should) that alter the state of something on an event trigger that feels like a hyperparameter tuning.

Reply:
Agree constant related to training is definitely going to be stored as a constant in the config file, but even with those you explicitly say where they end up in code. Logging doesn't alter state, so it should just be in a config? True, we should have an explicit decision on this.